library/core/class.oauth2.php

```diff
@@ -1045,4 +1079,156 @@ protected function decodeState(string $state): array {
             }
         }
     }
+
+    /**
+     * Exchange an OAuth access token for a Vanilla access token.
+     *
+     * @param TokensApiController $sender
+     * @param array $body
+     * @return \Garden\Web\Data
+     */
+    final public function tokensApiController_post_oauth(TokensApiController $sender, array $body): \Garden\Web\Data {
+        $sender->permission(Permissions::BAN_CSRF);
+
+        $in = $sender->schema([
+            'clientID:s',
+            'oauthAccessToken:s',
+        ], 'in');
+
+        $valid = $in->validate($body);
+
+        // Look up the specific addon that owns this client ID.
+        $instance = $this->getInstanceFromClientID($valid['clientID']);
+
+        try {
+            $result = $instance->issueAccessToken($valid['clientID'], $valid['oauthAccessToken']);
+        } catch (ContainerExceptionInterface $ex) {
+            throw new ServerException("There was an error getting the OAuth client instance.");
+        }
+
+        return new \Garden\Web\Data($result);
+    }
+
+    /**
+     * Connect the user payload with a user.
+     *
+     * @param array $payload The user profile.
+     * @return int
+     * @throws ClientException Throws an exception if the user cannot be connected for some reason.
+     * @throws Garden\Schema\ValidationException Throws an exception if the payload doesn't contain the required fields.
+     */
+    final private function sso(array $payload): int {
+        unset($payload['UserID']); // safety precaution due to Gdn_UserModel::connect() behaviour
+
+        /* @var \UserModel $userModel */
+        $userModel = $this->container->get(\UserModel::class);
+
+        $userID = $userModel->connect(
+            $payload['UniqueID'] ?? '',
+            static::PROVIDER_KEY,
+            $payload,
```

Comment:
This const does not exist in the Oauth2 class. The provider key was never a constant in this context because we hoped to have multiple OAuth2 connections at the same time.

Reply:
Great catch!